File: ToolKit/src/main/java/com/uday/android/util/ApkListData.java
package com.uday.android.util;
import android.content.pm.*;
import android.graphics.*;
import android.graphics.drawable.*;
import android.util.*;
import java.io.*;
import com.uday.android.toolkit.*;
import android.content.*;

public class ApkListData
{
	public File apkFile;
	public String PATH;
	public String NAME;
	public String VERSION_NAME="Invalid apk file";
	public int VERSION_CODE=0;
	public Drawable ICON;
	public boolean isSelectable=false,isSelected=false;
	public String SIZE="0kB";
	public String PACKAGE_NAME="parse error..!";
	public int titleColor=Color.BLACK;
	public boolean isInstalled=false,isOld=false,isInstalledVer=false;
	public String txtSearch;
	
	private PackageManager pm;
	private Context context;
	private PackageInfo pi;
	private OnAddedListener listener;

	public ApkListData(Context context,File apkFile,final PackageManager pm,Drawable icDefault){
		this.context=context;
		this.apkFile=apkFile;
		this.pm=pm;
		PATH=apkFile.getAbsolutePath();
		NAME=apkFile.getName();
		ICON=icDefault;
		SIZE=Utils.getSize(apkFile);
		
		
	}
	
	public ApkListData add(){
		((MainActivity)context).runInBackground(new Runnable(){
				@Override
				public void run(){
					try{
						pi=pm.getPackageArchiveInfo(PATH,0);
						pi.applicationInfo.sourceDir       = PATH;
						pi.applicationInfo.publicSourceDir = PATH;
						NAME=pi.applicationInfo.loadLabel(pm).toString();
						VERSION_NAME=pi.versionName;
						VERSION_CODE=pi.versionCode;
						ICON=pi.applicationInfo.loadIcon(pm);
						PACKAGE_NAME=pi.packageName;
						if(!PACKAGE_NAME.equals("com.uday.android.toolkit"))
							isSelectable=true;
						ApplicationInfo info = pm.getApplicationInfo(  
							PACKAGE_NAME, PackageManager.GET_UNINSTALLED_PACKAGES);
						isInstalled=true;
						Log.d(MainActivity.TAG,  "checkAppInstalledByName: "+PACKAGE_NAME+" : found");
						PackageInfo pInfo=pm.getPackageInfo(PACKAGE_NAME,0);
						if(pInfo.versionCode==VERSION_CODE){
							isInstalledVer=true;
						}else if(pInfo.versionCode>VERSION_CODE) isOld=true;
						
					}catch(Exception ex){
						//	Log.d(MainActivity.TAG,ex.toString());
						//	Log.d(MainActivity.TAG,  "checkAppInstalledByName:"+PACKAGE_NAME+" not found");
						isInstalled=false;
					}
					if(listener!=null){
						((MainActivity)context).runOnUiThread(new Runnable(){
							@Override
							public void run(){
								listener.onAdded();
							}
						});
					}
				}
			});
		return this;
	}
	
	public ApkListData setOnAddedListener(OnAddedListener listener){
		this.listener=listener;
		return this;
	}
	
	
	public static interface OnAddedListener{
		public void onAdded();
	}
}

File: app/src/main/java/com/uday/android/util/ApkListData.kt
package com.uday.android.util

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.util.Log

import com.uday.android.toolkit.MainActivity

import java.io.File


class ApkListData(
    private val context: Context,
    var apkFile: File,
    private val pm: PackageManager,
    var ICON: Drawable
) {
    var path: String = apkFile.absolutePath
    var name: String
    var versionName = "Invalid apk file"
    var versionCode = 0
    var isSelectable = false
    var isSelected = false
    var size: String
    var packageName = "parse error..!"
    var titleColor = MainActivity.colorOnBackground
    var isInstalled = false
    var isOld = false
    var isInstalledVer = false
    var txtSearch: String? = null
    var sizeLong: Long = 0
    private var pi: PackageInfo? = null
    private var listener: OnAddedListener? = null

    init {
        name = apkFile.name
        sizeLong = apkFile.length()
        size = Utils.getConventionalSize(sizeLong)
    }

    fun add(): ApkListData {
        (context as MainActivity).runInBackground(Runnable {
            try {
                pi = pm.getPackageArchiveInfo(path, 0)
                pi!!.applicationInfo.sourceDir = path
                pi!!.applicationInfo.publicSourceDir = path
                name = pi!!.applicationInfo.loadLabel(pm).toString()
                versionName = pi!!.versionName
                @Suppress("DEPRECATION")
                versionCode = pi!!.versionCode
                ICON = pi!!.applicationInfo.loadIcon(pm)
                packageName = pi!!.packageName
                if (packageName != "com.uday.android.toolkit")
                    isSelectable = true
//                val info = pm.getApplicationInfo(packageName, PackageManager.GET_UNINSTALLED_PACKAGES)
                isInstalled = true
                Log.d(MainActivity.TAG, "checkAppInstalledByName: $packageName : found")
                val pInfo = pm.getPackageInfo(packageName, 0)
                @Suppress("DEPRECATION")
                if (pInfo.versionCode == versionCode) {
                    isInstalledVer = true
                } else if (pInfo.versionCode > versionCode) isOld = true

            } catch (ex: Exception) {
                isInstalled = false
            }

            if (listener != null) {
                context.runOnUiThread { listener!!.onAdded() }
            }
        })
        return this
    }

    fun setOnAddedListener(listener: OnAddedListener): ApkListData {
        this.listener = listener
        return this
    }


    interface OnAddedListener {
        fun onAdded()
    }
}
